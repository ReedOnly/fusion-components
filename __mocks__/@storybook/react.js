import renderer from 'react-test-renderer';
import addons, { mockChannel } from '@storybook/addons';
import { withFusionContext } from '../../.storybook/withFusionContext';

addons.setChannel(mockChannel());

export const action = (actionName) => jest.fn();
const fusionContext = withFusionContext();
export const storiesOf = (groupName) => {
    const api = {
        decorators: [],
        parameters: {},

        add(storyName, story) {
            if (api.parameters.skipTest) {
                return;
            }

            describe(groupName, () => {
                it(storyName, () => {
                    // const decoratedStory = api.decorators.reduce((s, decorator) => {
                    //     try {
                    //         return decorator(s, { parameters: api.parameters });
                    //     } catch (e) {
                    //         return s;
                    //     }
                    // }, story);

                    const component = renderer.create(fusionContext(story));

                    // @TODO Enable custom names once released:
                    // > https://github.com/facebook/jest/pull/2094
                    expect(component.toJSON())
                        .toMatchSnapshot
                        // `${groupName}.${storyName}`
                        ();
                });
            });

            return api;
        },

        addDecorator() {
            return api;
        },

        addParameters(parameters) {
            api.parameters = { ...api.parameters, ...parameters };
            return api;
        },
    };

    return api;
};

export const addDecorator = () => {};
