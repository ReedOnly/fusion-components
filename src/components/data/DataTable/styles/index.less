.container {
    &.comfortable {
    }

    &.compact {
    }
    
    &.large {
    }

    &.medium {
    }

    &.small {
    }
}