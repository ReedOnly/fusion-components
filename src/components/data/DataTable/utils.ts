import React, { useState, useEffect } from 'react';
import {
    DataItemPropertyAccessor,
    DataItemPropertyAccessorFunction,
    DataTableColumn,
    DataItemBooleanAccessor,
    DataItemBooleanAccessorFunction,
} from './dataTableTypes';
import { useEventListener } from '@equinor/fusion-components';

export const getString = <T>(item: T, accessor: DataItemPropertyAccessor<T>): string => {
    if (typeof accessor === 'string') {
        return (item[accessor] as any).toString();
    }

    const accessorFunction = accessor as DataItemPropertyAccessorFunction<T>;

    return accessorFunction(item);
};

export const getBoolean = <T>(item: T, accessor: DataItemBooleanAccessor<T>): boolean => {
    if (typeof accessor === 'boolean') {
        return accessor;
    }

    const accessorFunction = accessor as DataItemBooleanAccessorFunction<T>;

    return accessorFunction(item);
};

export const getCellContent = <T>(
    item: T,
    column: DataTableColumn<T>,
    rowIndex: number
): React.ReactElement | string => {
    const CustomComponent = column.component;
    if (CustomComponent) {
        return React.createElement(CustomComponent, { item, rowIndex });
    }

    return getString(item, column.accessor);
};

const toCssUnit = (value: number | string) => {
    if (typeof value === 'number') {
        return value + 'px';
    }

    return value;
};

export const generateColumnTemplate = <T>(columns: DataTableColumn<T>[]) =>
    'max-content ' +
    columns.map(c => toCssUnit(`minmax(max-content, ${c.width || 'auto'})`)).join(' ');

const rowTemplate = 'calc(var(--grid-unit) * var(--row-height-multiplier))';
export const generateRowTemplate = <T>(rows: T[], expandedRows: T[], skeletonRows: number) => {
    if (!rows.length) {
        // Skeleton rows + header
        return `repeat(${skeletonRows + 1}, ${rowTemplate})`;
    }

    return (
        rowTemplate +
        ' ' +
        rows
            .map(row => {
                const isExpanded = expandedRows.findIndex(r => r === row) > -1;

                if (!isExpanded) {
                    return rowTemplate;
                }

                return `${rowTemplate} auto`;
            })
            .join(' ')
    );
};

const hasNarrowerParent = (node: HTMLElement, width?: number): boolean => {
    width = width || node.offsetWidth;
    const parent = node.parentElement;

    if (!parent) {
        return false;
    }

    if (parent.offsetWidth < width) {
        return true;
    }

    return hasNarrowerParent(parent, width);
};

/**
 * Calculates which columns can fit (and be visible) in a table.
 * It works by reacting to the resize event on the window
 * and checking if the current table is to wide for its parent.
 * @param columns All columns
 * @param ref A ref to the table wrapper
 */
export const useVisibleColumns = <T>(
    columns: DataTableColumn<T>[],
    ref: React.RefObject<HTMLDivElement>,
    deps: any[]
) => {
    const [collapsedColumns, setCollapsedColumns] = useState<DataTableColumn<T>[]>([]);
    const [testCollapsedColumns, setTestCollapsedColumns] = useState<DataTableColumn<T>[]>([]);
    const [shouldRecalculateColumns, setShouldRecalculateColumns] = useState(false);
    const [shouldTestColumns, setShouldTestColumns] = useState(false);
    const [resize, setResize] = useState({ from: 0, to: 0 });

    useEffect(() => {
        // We don't need to recalculate columns
        if (!shouldRecalculateColumns) {
            return;
        }

        // Remove already collapsed columns
        // Sort the columns by priority
        const sortedColumns = columns
            .filter(c => !collapsedColumns.find(cc => cc.key === c.key))
            .sort((a, b) => {
                if (!a.priority) {
                    return -1;
                } else if (!b.priority) {
                    return 1;
                }

                return b.priority - a.priority;
            });

        // Add the first of the sorted columns (worst priority) to the collapsed columns
        setCollapsedColumns([...collapsedColumns, sortedColumns[0]]);
    }, [shouldRecalculateColumns]);

    useEffect(() => {
        // We just collapsed a column, no need to continue collapsing until the checkParentWidth functions says so
        setShouldRecalculateColumns(false);
    }, [collapsedColumns]);

    const checkParentWidth = () => {
        if (!ref.current || shouldRecalculateColumns) {
            return;
        }

        // Check if the parent is narrower than the table
        const isNarrower = hasNarrowerParent(ref.current);

        const columnsLeft = columns.length - collapsedColumns.length;

        if (
            isNarrower &&
            (resize.from > resize.to || resize.from === resize.to) &&
            columnsLeft > 2
        ) {
            // The table is wider that the parent, and the user is making the window smaller
            // so we need to recalculate columns
            setShouldRecalculateColumns(true);
        } else if (!isNarrower && collapsedColumns.length && resize.from < resize.to) {
            // The table is not wider than the parend, the user is making the window bigger
            // and we have collapsed a couple of columns
            // so we test if we can expand one column without while keeping the table narrower than the parent
            const newCollapsedColumns =
                collapsedColumns.length > 1 ? collapsedColumns.slice(0, -1) : [];
            setTestCollapsedColumns(newCollapsedColumns);
            setShouldTestColumns(true);
        }
    };

    useEffect(() => {
        if (!ref.current || !shouldTestColumns) {
            return;
        }

        // Lets test if it worked with one less expanded column
        const isNarrower = hasNarrowerParent(ref.current);

        if (isNarrower) {
            // Nope!
            setTestCollapsedColumns([]);
        } else {
            // Yes! Apply the test to the actual collapsed columns
            setCollapsedColumns(testCollapsedColumns);
            setTestCollapsedColumns([]);
        }

        // We're done testing for now
        setShouldTestColumns(false);
    }, [testCollapsedColumns, shouldTestColumns, ref.current]);

    useEffect(checkParentWidth, [
        columns,
        ref.current,
        shouldRecalculateColumns,
        collapsedColumns,
        resize,
        ...deps,
    ]);

    const onResize = () => {
        if (!ref.current) {
            return;
        }

        const from = resize.to;
        const to = window.innerWidth;
        setResize({ from, to });
    };

    useEventListener(window, 'resize', onResize, [resize]);

    // Remove the collapsed columns, or the columns we're testing from all the columns before returning
    const visibleColumns = columns.filter(
        c =>
            !(testCollapsedColumns.length ? testCollapsedColumns : collapsedColumns).find(
                cc => cc.key === c.key
            )
    );

    return {
        visibleColumns,
        collapsedColumns,
    };
};
