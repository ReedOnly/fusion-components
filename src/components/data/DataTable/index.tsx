import * as React from 'react';

import useHorizontalBreakpoint, {
    HorizontalBreakpoint,
} from '../../../hooks/useHorizontalBreakpoint';

type DataTableProps = {
    test: string,
};

const sizeBreakpoints: HorizontalBreakpoint[] = [
    { key: 'large', width: 1080 },
    { key: 'medium', width: 767 },
    { key: 'small', width: 0 },
];

const DataTable: React.FC<DataTableProps> = (props: DataTableProps): React.ReactElement => {
    const [ref, breakpoint] = useHorizontalBreakpoint(sizeBreakpoints);
    console.log('breakpoint', breakpoint);

    
    return <div ref={ref}>{props.test}</div>;
};

export default DataTable;
