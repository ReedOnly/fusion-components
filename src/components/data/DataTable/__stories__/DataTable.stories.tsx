import * as React from 'react';
import { storiesOf } from '@storybook/react';
import { withKnobs, text } from '@storybook/addon-knobs';

import DataTable from '../index';

const stories = storiesOf('Data|DataTable', module);
stories.addDecorator(withKnobs);

stories.add('Default', () => <DataTable test={text('Test', 'Test av knobs')} />);
