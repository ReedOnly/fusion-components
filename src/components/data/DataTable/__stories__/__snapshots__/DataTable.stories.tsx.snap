// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Data|Data Table Default 1`] = `
<div
  className="container comfortable"
>
  <div
    className="container comfortable"
  >
    <div
      className="container comfortable"
    />
  </div>
  <div
    className="overlay"
  />
  <div
    className="snacks"
  />
  <div
    className="dialog"
  />
</div>
`;

exports[`Data|Data Table Multi selectable 1`] = `
<div
  className="container comfortable"
>
  <div
    className="container comfortable"
  >
    <div
      className="container comfortable"
    />
  </div>
  <div
    className="overlay"
  />
  <div
    className="snacks"
  />
  <div
    className="dialog"
  />
</div>
`;

exports[`Data|Data Table No Columns Collapse 1`] = `
<div
  className="container comfortable"
>
  <div
    className="container comfortable"
  >
    <div
      className="container comfortable"
    />
  </div>
  <div
    className="overlay"
  />
  <div
    className="snacks"
  />
  <div
    className="dialog"
  />
</div>
`;

exports[`Data|Data Table Single selectable 1`] = `
<div
  className="container comfortable"
>
  <div
    className="container comfortable"
  >
    <div
      className="container comfortable"
    />
  </div>
  <div
    className="overlay"
  />
  <div
    className="snacks"
  />
  <div
    className="dialog"
  />
</div>
`;

exports[`Data|Data Table With skeleton 1`] = `
<div
  className="container comfortable"
>
  <div
    className="container comfortable"
  >
    <div
      className="container comfortable"
    />
  </div>
  <div
    className="overlay"
  />
  <div
    className="snacks"
  />
  <div
    className="dialog"
  />
</div>
`;
